google/cloud/pubsub/publisher_connection.cc

```diff
@@ -138,13 +138,13 @@ std::shared_ptr<pubsub::PublisherConnection> MakePublisherConnection(
     if (options.message_ordering()) {
       auto factory = [topic, options, sink, cq](std::string const& key) {
         return BatchingPublisherConnection::Create(
-            topic, options, key, SequentialBatchSink::Create(std::move(sink)),
-            cq);
+            std::move(topic), std::move(options), key,
+            SequentialBatchSink::Create(std::move(sink)), std::move(cq));
       };
       return OrderingKeyPublisherConnection::Create(std::move(factory));
     }
     return RejectsWithOrderingKey::Create(BatchingPublisherConnection::Create(
-        std::move(topic), std::move(options), {}, sink, std::move(cq)));
+        topic, options, {}, std::move(sink), std::move(cq)));
   };
   auto connection = make_connection();
```

Comment:
I am beginning to think this is too clever for its own good.

Maybe we should do this (starting around line 133):


```cc
auto cp = background->cq();
auto sink = DefaultBatchSink::Create(stub, cq, std::move(retry_policy), std::move(backoff_policy));
auto connection = RejectsWithOrderingKey::Create(BatchingPublisherConnection::Create(topic, options, {}, sink, cq);
if (options.message_ordering()) {
  auto factory = [topic, options, sink, cq](std::string const& key) {
    return BatchingPublisherConnection::Create(topic, options, key, SequentialBatchSink(std::move(sink), std::move(cq)));
  };
  connection = OrderKeyPublisherConnection(std::move(factory));
}
```

It makes a few more copies than it should, but it is probably more readable.  Alternatively, moving the lambda to a function may make it more obvious.  Your call.

Reply:
I think the code is fine. I like that `BatchingPublisherConnection::Create()` is only called once